src/main/java/org/jitsi/jibri/capture/ffmpeg/executor/impl/AbstractFfmpegExecutor.kt

```diff
@@ -0,0 +1,120 @@
+        ffmpegOutputLogger.addHandler(FfmpegFileHandler())
+    }
+
+    /**
+     * Get the shell command to use to launch ffmpeg
+     */
+    protected abstract fun getFfmpegCommand(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink): String
+
+    override fun launchFfmpeg(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink) {
+        processBuilder.command(getFfmpegCommand(ffmpegExecutorParams, sink).split(" "))
+        processBuilder.redirectErrorStream(true)
+        logger.info("Running ffmpeg command:\n ${processBuilder.command()}")
+        currentFfmpegProc = processBuilder.start()
+        // Tee ffmpeg's output so that we can analyze its status and log everything
+        ffmpegOutputTee = Tee(currentFfmpegProc!!.inputStream)
+        ffmpegTail = Tail(ffmpegOutputTee!!.addBranch())
+        // Read from a tee branch and log to a file
+        executor.submit {
+            val reader = BufferedReader(InputStreamReader(ffmpegOutputTee!!.addBranch()))
+            while (true) {
+                ffmpegOutputLogger.info(reader.readLine())
+            }
+        }
+
+        logger.debug("Launched ffmpeg, is it alive? ${currentFfmpegProc?.isAlive}")
+    }
+
+    override fun getExitCode(): Int? {
+        if (currentFfmpegProc?.isAlive == true) {
+            return null
+        }
+        return currentFfmpegProc?.exitValue()
+    }
+
+    override fun isHealthy(): Boolean {
+        //TODO: should we only consider 2 sequential instances of not getting a frame=
+        // encoding line "unhealthy"?
+        currentFfmpegProc?.let {
+            val ffmpegOutput = ffmpegTail?.mostRecentLine ?: ""
+            val parsedOutputLine = OutputParser().parse(ffmpegOutput)
+            if (!it.isAlive) {
+                logger.error("Ffmpeg is no longer running, its most recent output line was: $ffmpegOutput")
+                return false
+            }
+
+            if (!parsedOutputLine.containsKey("frame")) {
+                logger.error("Ffmpeg is running but doesn't appear to be encoding.  " +
+                        "Its most recent output line was $ffmpegOutput")
+                return false
+            } else {
+                logger.debug("Ffmpeg appears healthy: $parsedOutputLine")
+                return true
+            }
+        }
+        return false
+    }
+
+    override fun stopFfmpeg() {
+        executor.shutdown()
+        currentFfmpegProc?.let {
+            val pid = pid(it)
+            logger.info("Sending SIGINT to ffmpeg proc $pid")
+            Runtime.getRuntime().exec("kill -s SIGINT $pid")
+        } ?: run {
+            logger.info("Ffmpeg had already exited")
+        }
+        currentFfmpegProc?.waitFor(10, TimeUnit.SECONDS)
+        currentFfmpegProc?.isAlive.let {
+            // This isn't great, as killing ffmpeg this way will corrupt
+            // the entire recording (from what I've seen)
+            currentFfmpegProc?.destroyForcibly()
```

Comment:
Then what is the alternative to forcibly killing? Is 10 seconds an informed number?

Reply:
i don't know of any alternative :( and 10 is not terribly significant, just seemed like a reasonable value.